enhancements/authentication/custom-scc-preemption-prevention.md

```diff
@@ -0,0 +1,260 @@
+additional custom SCCs on top of the default out of the box SCCs.
+Custom SCCs are not considered different from the perspective of
+the SCC admission plugin.
+
+When a pod is being admitted, the admission plugin lists all SCCs available in the system, including all custom SCCs.
+The list of SCCs is then sorted by:
+- SCC priority in descending order
+- SCCs with the same priority: SCC restriction score in descending order (most restrictive first, least restrictive last)
+- SCCs with the same restriction score: lexical ordering of the SCC name in descending order
+
+An SCC is considered applicable if the requesting user or the
+pod's service account has `use` permissions against the `security.openshift.io.securitycontextconstraints` resource
+for an SCC in the requested namespace.
+If an SCC is applicable, the pod's manifest is potentially mutated according to the SCC logic.
+
+One can see that custom SCCs can easily overrule existing out of the box SCCs
+if the pod's service account or the requesting user has sufficient privileges to use any of them.
+
+In that case, if a custom SCC has higher priority, is more restrictive, or it's name is lexically ordered before an out of the box SCC,
+any OpenShift out of the box core pod will be mutated according to the custom SCC specified behavior.
+
+This can lead to unknown mutations of core OpenShift workloads.
+Worst case it can render OpenShift core workloads dysfunctional
+if i.e. a custom SCC is more restrictive than a previously applicable less restrictive out of the box SCC that was applicable to that workload.
+
+This proposal includes two suggested protection modes, a preferred "hard protection" and a "soft protection" variant
+which are described below.
+
+We need a mechanism to protect OpenShift core workloads such that they are not
+subject to pod mutation due to user provided custom SCCs.
+
+### User Stories
+
+1. As an OpenShift core developer I want to make sure that my shipped workload
+is guaranteed to be associated with an out of the box SCC.
+
+2. As an OpenShift core developer I want to make sure that my shipped workload
+is not preempted by custom user provided SCCs.
+
+### Goals
+
+1. Protect core OpenShift workloads from being preempted by user provided custom SCCs.
+
+2. Be able to pin core OpenShift workloads to a concrete out of the box SCC.
+
+3. Make this change generally available for end users.
+
+### Non-Goals
+
+## Proposal
+
+### "hard protection" - Requesting a required SCC
+
+This proposal suggests to implement a mechanism which enables OpenShift core workloads to enforce
+an SCC to be pinned to a workload.
+This is accomplished by setting an annotation named `openshift.io/required-scc`
+whose value contains the name of the SCC that is supposed to be pinned.
+
+In order to prevent accidental changes to that annotation,
+admission will fail admission if the `openshift.io/required-scc is being changed or added
+for update events.
```

Comment:
assuming i'm correct in https://github.com/openshift/enhancements/pull/1391/files#r1262948327

then i see no reason to fail admission when changing this annotation, except when changing it on pods themselves.  Is that what was implemented?

basically if you change it on the pod template of a deployment, i'd then expect the newly created pods to use the new SCC, or fail to use it and be rejected.

Reply:
> except when changing it on pods themselves. Is that what was implemented?

correct, and this is being implemented. Hence, if the annotation is changed as part of the deployment's pod template specification it will work as old pods are being deleted and new pods are being created.

I clarified this in the sentence to make it more clear.